### PR TITLE
[IA-3268] list azure runtime

### PR DESCRIPTION
**Pull request by jdcanas**
https://broadworkbench.atlassian.net/browse/IA-3268

Implement listRuntimeV2.

For the most part, this list runtime is fully functional and non-azure specific. As such, a lot of the lines here are duplicated runtimeV1 test code so that when we start switching over we have the same level of test coverage.

### CODE REVIEW

File: http/src/main/resources/swagger/api-docs.yaml [hunk 1]
```diff
@@ -388,7 +388,7 @@ paths:
             application/json:
               schema:
                 $ref: "#/components/schemas/ErrorReport"
-  /api/google/v1/runtimes/{googleProject}/{name}/start:
+  "/api/google/v1/runtimes/{googleProject}/{name}/start":
```
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml:391]: I don't think the quotes are necessary?

File: http/src/main/resources/swagger/api-docs.yaml [hunk 2]
```diff
@@ -432,7 +432,192 @@ paths:
...
+            type: string
+      responses:
+        "200":
+          description: List of runtimes
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: "#/components/schemas/ListRuntimeResponse"
```
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml:543]: should `ListRuntimeResponse` be a different one?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/WsmDao.scala [hunk 3]
```diff
@@ -277,12 +285,16 @@ object WsmDecoders {
...
+
   implicit val getWorkspaceResponseDecoder: Decoder[WorkspaceDescription] = Decoder.instance { c =>
     for {
       id <- c.downField("id").as[UUID]
       displayName <- c.downField("displayName").as[String]
-      azureContext <- c.downField("azureContext").as[AzureCloudContext]
-    } yield WorkspaceDescription(WorkspaceId(id), displayName, azureContext)
+      azureContext <- c.downField("azureContext").as[Option[AzureCloudContext]]
+      gcpContext <- c.downField("gcpContext").as[Option[WsmGcpContext]]
+    } yield WorkspaceDescription(WorkspaceId(id), displayName, azureContext, gcpContext.map(_.projectId))
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/WsmDao.scala]: can we have a `Decoder[WorkspaceId]`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala [hunk 4]
```diff
@@ -179,27 +204,75 @@ class AzureServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
...
+          for {
+            uuid <- Either.catchNonFatal(UUID.fromString(r.samResource.resourceId))
+          } yield WsmResourceSamResourceId(WsmControlledResourceId(uuid))
+        }
+        .flatMap(either =>
+          either match {
+            case Right(id) => List(id)
+            case Left(_)   => List.empty
+          }
+        )
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: this means /api/v2/runtimes won't show runtimes whose sam resource Id is not UUID. Can we make sure that in prod, all active runtimes do have UUID as Sam resource ID?

**Comment by jdcanas** [comment]
jenkis retest, fiab-start failure

**Comment by jdcanas** [comment]
jenkins retest, fiab-start failure due to sam

File: http/src/main/resources/swagger/api-docs.yaml (hunk 1 above)
Comment (jdcanas) [reply http/src/main/resources/swagger/api-docs.yaml:391]: They aren't but the majority of our routes have them in this file, so i added for consistency

File: http/src/main/resources/swagger/api-docs.yaml (hunk 2 above)
Comment (jdcanas) [reply http/src/main/resources/swagger/api-docs.yaml:543]: It is the same json entity. Trying to keep it consistent (the underlying runtime model is the same and all) so its easier for the consumer.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala (hunk 4 above)
Comment (jdcanas) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: Hmm... that doesn't seem to be the case.

I'm actually realizing this logic is pretty bad as it assumes all runtimes in the future will be registered with WSM, which isn't the case. Unfortunately, I think we will need 2 sam calls for this to:
1.  support azure runtimes registered as a WSM controlled resource
2. a fallback to check all non-creator-is-caller in sam the way the current runtime APIs do (as we have no plans to migrate google runtimes to be registered in WSM) 

I'll push up the code shortly

**Comment by Qi77Qi** [comment]
one thing I notice is `checkSamPermission` is used in `deleteRuntime` as well...we should probably check `writer` permission for delete, but `checkSamPermission` is only checking `reader`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/RuntimeServiceDbQueries.scala [hunk 5]
```diff
@@ -167,7 +166,37 @@ object RuntimeServiceDbQueries {
         .filter(_.cloudContextDb === p.asCloudContextDb)
         .filter(_.cloudProvider === p.cloudProvider)
     )
-    val runtimeQueryJoinedWithLabel = runtimeLabelQuery(clusterQueryFilteredByProject)
+
+    joinAndFilterByLabelForList(labelMap, clusterQueryFilteredByProject)
+  }
+
+  def listRuntimes(labelMap: LabelMap,
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/RuntimeServiceDbQueries.scala]: can we give this one a different name...overloading functions is generally discouraged in Scala

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala [hunk 6]
```diff
@@ -179,27 +202,79 @@ class AzureServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
...
+              case Left(_)   => List.empty
+            }
+          } yield WsmResourceSamResourceId(WsmControlledResourceId(uuid))
+        }
+
+      samVisibleWsmControlledResourceSamIds <- NonEmptyList
+        .fromList(wsmControlledResourceSamIds)
+        .fold[F[List[WsmResourceSamResourceId]]](F.pure(List.empty)) { ids =>
+          authProvider.filterUserVisible(ids, userInfo)
+        }
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: can do this instead

```
NonEmptyList
        .fromList(wsmControlledResourceSamIds)
        .traverse { ids =>
          authProvider.filterUserVisible(ids, userInfo)
        }
```

Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: can do this instead

```
NonEmptyList
        .fromList(wsmControlledResourceSamIds)
        .traverse { ids =>
          authProvider.filterUserVisible(ids, userInfo)
        }
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala [hunk 7]
```diff
@@ -179,27 +202,79 @@ class AzureServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
...
+
+      // We must also check the RuntimeSamResourceId in sam to support already existing and newly created google runtimes
+      samVisibleRuntimeSamResourceIds <- NonEmptyList
+        .fromList(runtimesUserIsNotCreator.map(_.samResource))
+        .fold[F[List[RuntimeSamResourceId]]](F.pure(List.empty)) { ids =>
+          authProvider.filterUserVisible(ids, userInfo)
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: does this still have equivalent of project level fallback? It doesn't like this logic do anything special if the caller is workspace owner?

Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: does this still have equivalent of project level fallback? It doesn't like this logic do anything special if the caller is workspace owner?

Comment (jdcanas) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: We do check if the caller is the creator of the runtime, but I'm not sure exactly what a 'project fallback' entails from the code. Does it check the owner of the project? 

So the corollary would be to check if the user is he owner of the workspace in addition to the sam runtime permission we are checking now for this use case?

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: sry just saw the comment,

>So the corollary would be to check if the user is he owner of the workspace in addition to the sam runtime permission we are checking now for this use case?

yes!

Comment (jdcanas) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala]: cc @Qi77Qi

**Review by Qi77Qi — APPROVED** [review]
lgtm..would be good to verify things on Dev

**Pull request merged**